### PR TITLE
KOGITO-6698: [DMN Editor] input and select boxes in header cells popups errors with esc key

### PR DESCRIPTION
**Pull request by fantonangeli**
https://issues.redhat.com/browse/KOGITO-6698

Fixes:

- select boxes in header cells popups, user can cancel context menu of header cell by pressing escape, however if an selectbox selection was changed, before, such change is not canceled

- input boxes in header cells popups, if user change column name of a decision table input column, then he press escape, then column is correctly not renamed, however if the menu is reopened, the new value is stored there

### CODE REVIEW

File: packages/boxed-expression-component/it-tests/integration/popover.spec.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as buildEnv from "@kie-tools/build-env";
+
+describe("Context Expression Tests", () => {
```
Comment (vpellegrino) [inline packages/boxed-expression-component/it-tests/integration/popover.spec.ts]: There is something in the description of these tests that doesn't sound good to me:
"**should** user **can** ...", please review them

File: packages/boxed-expression-component/src/components/PopoverMenu/PopoverMenu.tsx [hunk 2]
```diff
@@ -35,6 +36,14 @@ export interface PopoverMenuProps {
   hasAutoWidth?: boolean;
   /** Popover min width */
   minWidth?: string;
+  /**
+   * Lifecycle function invoked when the popover has fully transitioned out.
```
Comment (vpellegrino) [inline packages/boxed-expression-component/src/components/PopoverMenu/PopoverMenu.tsx]: can you please give more details about `onHide` and `onCancel`? Looking at the comments, they look the same

File: packages/boxed-expression-component/src/components/PopoverMenu/PopoverMenu.tsx [hunk 3]
```diff
@@ -46,7 +55,36 @@ export const PopoverMenu: React.FunctionComponent<PopoverMenuProps> = ({
   className,
   hasAutoWidth,
   minWidth,
+  onHide = () => {},
+  onCancel = () => {},
 }: PopoverMenuProps) => {
+  const [isVisible, setIsVisible] = useState(false);
+
+  const shouldOpen = useCallback((showFunction?: () => void, event?: MouseEvent | KeyboardEvent) => {
+    // setIsVisible(true);
```
Comment (vpellegrino) [inline packages/boxed-expression-component/src/components/PopoverMenu/PopoverMenu.tsx]: is this comment wanted?

File: packages/boxed-expression-component/src/components/PopoverMenu/PopoverMenu.tsx [hunk 4]
```diff
@@ -46,7 +55,36 @@ export const PopoverMenu: React.FunctionComponent<PopoverMenuProps> = ({
...
+  }, []);
+
+  const shouldClose = useCallback(
+    (_tip, hideFunction?: () => void, event?: MouseEvent | KeyboardEvent) => {
+      // if the esc key has been pressed with a Select component open
+      if ((event?.target as Element).closest(".pf-c-select__menu")) {
+        return;
+      }
+
+      // setIsVisible(false);
```
Comment (vpellegrino) [inline packages/boxed-expression-component/src/components/PopoverMenu/PopoverMenu.tsx]: is this comment wanted?

File: packages/boxed-expression-component/src/components/PopoverMenu/PopoverMenu.tsx [hunk 5]
```diff
@@ -46,7 +55,36 @@ export const PopoverMenu: React.FunctionComponent<PopoverMenuProps> = ({
...
+
+      if (event instanceof KeyboardEvent && /^esc.*/i.test(event?.key)) {
```
Comment (vpellegrino) [inline packages/boxed-expression-component/src/components/PopoverMenu/PopoverMenu.tsx]: I suggest wrapping the `test` check in a variable with a meaningful name

File: packages/boxed-expression-component/src/components/PopoverMenu/PopoverMenu.tsx [hunk 6]
```diff
@@ -46,7 +55,36 @@ export const PopoverMenu: React.FunctionComponent<PopoverMenuProps> = ({
...
+        onCancel(event);
+      } else {
+        onHide();
+      }
+
+      hideFunction && hideFunction();
```
Comment (vpellegrino) [inline packages/boxed-expression-component/src/components/PopoverMenu/PopoverMenu.tsx]: you could just use `hideFunction?.()`

**Review by vpellegrino — CHANGES_REQUESTED** [review]
Just a few suggestions from a code perspective. I am going now to checkout your branch and let you know if I found anything strange while manual testing it.
Thanks, @fantonangeli

File: packages/boxed-expression-component/it-tests/integration/popover.spec.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
```
Comment (jomarko) [inline packages/boxed-expression-component/it-tests/integration/popover.spec.ts]: 2022

File: packages/boxed-expression-component/it-tests/integration/popover.spec.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  });
+
+  it("should user can cancel edit of data type in context menu of header cell by pressing escape", () => {
+    // open the context menu
+    cy.contains("th", "Expression Name").as("ExpressionNameCell").click();
+
+    // open the dataType select
+    cy.ouiaId("edit-expression-data-type")
+      .as("dataTypeInput")
+      .find("span.pf-c-select__toggle-text")
```
Comment (jomarko) [inline packages/boxed-expression-component/it-tests/integration/popover.spec.ts]:
```suggestion
      .find("button")
```

File: packages/boxed-expression-component/it-tests/integration/popover.spec.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      .click({ force: true });
+
+    // select "context" type
+    cy.get("@dataTypeInput").contains("context").click({ force: true });
+
+    // context menu should be open. This check is to check that the context menu is not closed after data type selection
+    cy.ouiaId("expression-popover-menu").should("be.visible");
+
+    // Cancel the editing
+    cy.ouiaId("expression-popover-menu").type("{esc}").wait(500);
```
Comment (jomarko) [inline packages/boxed-expression-component/it-tests/integration/popover.spec.ts]:
```suggestion
    cy.ouiaId("expression-popover-menu").type("{esc}");
    cy.ouiaId("expression-popover-menu").should("not.exist");
```

File: packages/boxed-expression-component/it-tests/integration/popover.spec.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    cy.contains("th", "Expression Name").as("ExpressionNameCell").click();
+
+    // change the expression name
+    cy.get("#expression-name").type("{selectall}Newname", { force: true });
+
+    // Cancel the editing
+    cy.ouiaId("expression-popover-menu").contains("Edit Expression").click();
+
+    // context menu should be open. This check is to check that the context menu is not closed after data type selection
+    cy.ouiaId("expression-popover-menu").should("be.visible").type("{esc}").wait(500);
```
Comment (jomarko) [inline packages/boxed-expression-component/it-tests/integration/popover.spec.ts]:
```suggestion
    cy.ouiaId("expression-popover-menu").should("be.visible").type("{esc}");
    cy.ouiaId("expression-popover-menu").should("not.exist");
```

File: packages/boxed-expression-component/it-tests/integration/popover.spec.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    cy.contains("th", "Expression Name").as("ExpressionNameCell").click();
+
+    // expression name input should have original value
+    cy.get("#expression-name").should("have.value", "Expression Name");
+
+    // dataType select should have original value
+    cy.ouiaId("edit-expression-data-type").should("contain.text", "Undefined");
+
+    // close the context menu
+    cy.get("body").click();
```
Comment (jomarko) [inline packages/boxed-expression-component/it-tests/integration/popover.spec.ts]: I tend to think we do not need to close the context menu as the test does not do more actions. my point is to keep the test execution time as small as possible.

**Review by jomarko — COMMENTED** [review]
I agree with @vpellegrino thaht `describe` and `it` text annotations should be revisited.  I also put some comments inline.
- one comment to locator preference based just on element name instead of now human readable class names
- cypress suggest to not use wait if possible, I think we can just wait until popover element disappear. I tried locally and the suggested changes passed

However I haven't found any issue during manual check @fantonangeli . Thank you.

**Comment by fantonangeli** [comment]
Jenkins please retest this

**Comment by vpellegrino** [comment]
@fantonangeli can you please solve branch's conflicts?

**Review by vpellegrino — APPROVED** [review]
LGTM, thank you

**Review by jomarko — APPROVED** [review]

**Review by yesamer — APPROVED** [review]
@fantonangeli LGTM, thank you.

**Comment by yesamer** [comment]
Hi @ljmotta ! Can you please review this PR? Thx!

**Review by ljmotta — APPROVED** [review]
LGTM, thanks for this PR!

**Pull request merged**